lib/Sema/CSGen.cpp

```diff
@@ -3815,10 +3888,27 @@ namespace {
       }
 
       if (auto type = CG.visit(expr)) {
+
         auto simplifiedType = CS.simplifyType(type);
 
         CS.setType(expr, simplifiedType);
 
+        // If this is the end of an unresolved member chain, inject the
+        // appropriate constraints to link the contextual type to the (implicit)
+        // base of the chain.
+        if (isMemberChainTail(CG.getConstraintSystem(), expr)) {
+          auto *chainBase = getMemberChainBase(expr);
```

Comment:
Could we refactor `getMemberChainBase` to return `UnresolvedMemberExpr *` for a valid chain that would clarify it's nature here. Also, as I mentioned before, I think this could only be done once per expression - pre-walk could identify all of the viable chains and record their tails/heads to eliminate the need to do that for every sub-expression...

Reply:
> Could we refactor getMemberChainBase to return UnresolvedMemberExpr * for a valid chain that would clarify it's nature here.

Sure, that works for me.

> Also, as I mentioned before, I think this could only be done once per expression - pre-walk could identify all of the viable chains and record their tails/heads to eliminate the need to do that for every sub-expression...

I'm actually reworking this approach. Rewriting the AST in `ConstraintWalker` while also basing the rewrite logic on `ConstraintSystem::getParentExpr` causes some issues when parsing expressions that require recursive constraint regeneration (I think for a new solver scope, perhaps?). Instead, I am going to try to generate the `UnresolvedMemberChainResultExpr` wrapper inside `walkToExprPre` (using `ASTWalker::Parent` instead of `ConstraintSystem::getParentExpr`, and then generate the constraints inside `ConstraintGenerator` with a new `visitUnresolvedMemberChainResultExpr` method. Does that sound reasonable to you?